### ISSUE TITLE
Adding third `ABC` type to a loop causes false positive type error

### ISSUE THREAD

**Issue by johnthagen** [python/mypy#16509]
**Bug Report**

Adding a third `ABC` to a list of types in a `for` loop causes a Mypy type checking false positive error.

In the code below, if you _remove_ the second `C2` there is no error, but adding a third `ABC` to the iteration for some reason triggers this error.

**To Reproduce**

- https://mypy-play.net/?mypy=latest&python=3.10&gist=273ca66221a89cf15a8ff1c521a067e3

```python
from abc import ABC, abstractmethod


class Base(ABC):
    @abstractmethod
    def a(self) -> None:
        ...


class C1(Base):
    def a(self) -> None:
        return


class C2(Base):
    def a(self) -> None:
        return


for c in (C1, C2, C2):
    c()
```

**Expected Behavior**

No type check error.

**Actual Behavior**

Produces false positive type check error:

```
main.py:24: error: Cannot instantiate abstract class "Abstract" with abstract attribute "a"  [abstract]
```

**Your Environment**

- Mypy version used: 1.7.0
- Mypy command-line flags: N/A
- Mypy configuration options from `mypy.ini` (and other config files):

```toml
[tool.mypy]
strict = true
```

- Python version used: 3.10.11

**Comment by AlexWaygood** [comment]
I think this is just another instance of #3115?

**Comment by JelleZijlstra** [comment]
It's true though that mypy behaves differently depending on whether you iterate over `(C1, C2)` or `(C1, C2, C2)`. Here's a slightly expanded mypy play: https://mypy-play.net/?mypy=latest&python=3.10&gist=41c9f315bc22c5c19c40f60196042567

I'm not sure how to explain that change. reveal_type() doesn't change between the two cases.

**Comment by KotlinIsland** [comment]
If we change it to basedmypy we can see that it's working as expected
https://mypy-play.net/?mypy=basedmypy-latest&python=3.10&gist=41c9f315bc22c5c19c40f60196042567

**Comment by hauntsaninja** [comment]
Fixed in https://github.com/python/mypy/pull/17408
